browser/decentralized_dns/test/utils_unittest.cc

```diff
@@ -95,4 +97,40 @@ TEST_F(UtilsUnitTest, ResolveMethodMigration) {
   EXPECT_TRUE(IsENSResolveMethodAsk(local_state()));
 }
 
+TEST_F(UtilsUnitTest, Dot888Test) {
+  net::IPAddress ip_address;
+  // These tests were passing without .888 fix in `url_canon_ip.cc`.
+  EXPECT_TRUE(GURL("http://1.1.888").is_valid());
+  EXPECT_TRUE(ip_address.AssignFromIPLiteral("1.1.888"));
+  EXPECT_TRUE(GURL("http://123.888").is_valid());
+  EXPECT_TRUE(ip_address.AssignFromIPLiteral("123.888"));
+  EXPECT_TRUE(GURL("http://1.123.888").is_valid());
+  EXPECT_TRUE(ip_address.AssignFromIPLiteral("1.123.888"));
+  EXPECT_TRUE(GURL("http://.com").is_valid());
+  EXPECT_FALSE(ip_address.AssignFromIPLiteral(".com"));
+
+  // These tests start passing with .888 fix in `url_canon_ip.cc`.
+  const char* cases[] = {
+      "test.888",         // Non-ipv4 component case.
+      "test1.test2.888",  // Non-ipv4 component case.
+      "1.2.3.4.888",      // Too many components case.
```

Comment:
Please also add `1.2.3.888` to check that it's a valid URL but not a valid IPv4 URL.

Reply:
added in https://github.com/brave/brave-core/pull/14588/commits/a6ecf5cf8bd4056e6eed1f79202c4304a856b594
It actually didn't pass as I didn't cover residual bits case in parses(I wrongly assumed that was not .888 case)

Reply:
Now DoIPv4AddressToNumber never reports BROKEN when last component is .888

Reply:
> Now DoIPv4AddressToNumber never reports BROKEN when last component is .888

Is that the right behavior? I mean if you have `1.2.3.888`, which is not a valid IPv4 and try to convert this IPv4 URL to a number, it seems like it should fail because the input was not an IPv4 address in the first place, no?

Reply:
For correct IPv4 address parser should say that it is [exactly IPv4 address](https://source.chromium.org/chromium/chromium/src/+/main:net/base/ip_address.cc;l=136;drc=ba884f9cab8d36478acb75ad668280e05b61fb1c;bpv=1;bpt=1), so we don't break it
For Url parser now says it is [not a broken ipv4 address, but something else.](https://source.chromium.org/chromium/chromium/src/+/main:url/url_canon_ip.cc;l=222;drc=ba884f9cab8d36478acb75ad668280e05b61fb1c;bpv=1;bpt=1)

Reply:
I see, so we now have this?

- 1.2.3.4 -> valid URL, valid IPv4
- 1.2.3.889 -> invalid URL, invalid IPv4
- 1.2.3.888 -> valid URL, but not an IPv4 at all

It might be worth adding `1.2.3.889` as a test case too just to make sure that we don't change that one.

Reply:
added https://github.com/brave/brave-core/pull/14588/commits/692c0ffaffa78bab39f1a5d3bdfe5efe5a343e65
also now checking for 888 ending being an exact string to avoid 888 in hex or oct form.